File: LLD-main/LLD-main/Lecture 39/C++ Code/MementoPattern.cpp
#include <iostream>
#include <string>
#include <map>
#include <memory>

using namespace std;

// Memento - Stores database state snapshot
class DatabaseMemento {
private:
    map<string, string> data;
    
public:
    DatabaseMemento(const map<string, string>& dbData) {
        this->data = dbData;
    }
    
    map<string, string> getState() const {
        return data;
    }
};

// Originator - The database whose state we want to save/restore
class Database {
private:
    map<string, string> records;
    
public:
    // Insert a record
    void insert(const string& key, const string& value) {
        records[key] = value;
        cout << "Inserted: " << key << " = " << value << endl;
    }
    
    // Update a record
    void update(const string& key, const string& value) {
        if (records.find(key) != records.end()) {
            records[key] = value;
            cout << "Updated: " << key << " = " << value << endl;
        } else {
            cout << "Key not found for update: " << key << endl;
        }
    }
    
    // Delete a record
    void remove(const string& key) {
        auto it = records.find(key);
        if (it != records.end()) {
            records.erase(it);
            cout << "Deleted: " << key << endl;
        } else {
            cout << "Key not found for deletion: " << key << endl;
        }
    }
    
    // Create memento - Save current state
    DatabaseMemento* createMemento() {
        cout << "Creating database backup..." << endl;
        return new DatabaseMemento(records);
    }
    
    // Restore from memento - Rollback to saved state
    void restoreFromMemento(const DatabaseMemento& memento) {
        records = memento.getState();
        cout << "Database restored from backup!" << endl;
    }
    
    // Display current database state
    void displayRecords() {
        cout << "\n--- Current Database State ---" << endl;
        if (records.empty()) {
            cout << "Database is empty" << endl;
        } else {
            for (const auto& record : records) {
                cout << record.first << " = " << record.second << endl;
            }
        }
        cout << "-----------------------------\n" << endl;
    }
};

// Caretaker - Manages the memento (transaction manager)
class TransactionManager {
private:
    DatabaseMemento* backup;
    
public:
    TransactionManager() : backup(nullptr) {}
    
    // Destructor to clean up memory
    ~TransactionManager() {
        if (backup) {
            delete backup;
        }
    }
    
    // Begin transaction - create backup
    void beginTransaction(Database& db) {
        cout << "=== BEGIN TRANSACTION ===" << endl;
        if (backup) {
            delete backup; // Clean up previous backup
        }
        backup = db.createMemento();
    }
    
    // Commit transaction - discard backup
    void commitTransaction() {
        cout << "=== COMMIT TRANSACTION ===" << endl;
        if (backup) {
            delete backup;
            backup = nullptr;
        }
        cout << "Transaction committed successfully!" << endl;
    }
    
    // Rollback transaction - restore from backup
    void rollbackTransaction(Database& db) {
        cout << "=== ROLLBACK TRANSACTION ===" << endl;
        if (backup) {
            db.restoreFromMemento(*backup);
            delete backup;
            backup = nullptr;
            cout << "Transaction rolled back!" << endl;
        } 
        else {
            cout << "No backup available for rollback!" << endl;
        }
    }
};

int main() {
    Database db;
    TransactionManager txManager;
   
    //success scenario
    txManager.beginTransaction(db);
    db.insert("user1", "Aditya");
    db.insert("user2", "Rohit");
    txManager.commitTransaction();

    db.displayRecords();

    // Failed scenario
    txManager.beginTransaction(db);
    db.insert("user3", "Saurav");
    db.insert("user4", "Manish");

    db.displayRecords();
    
    // Some error -> Rollback
    cout << "ERROR: Something went wrong during transaction!" << endl;
    txManager.rollbackTransaction(db);
    
    db.displayRecords();
    
    return 0;
}

File: LLD-main/LLD-main/Lecture 39/Java Code/MementoPattern.java
import java.util.*;

// Memento - Stores database state snapshot
class DatabaseMemento {
   private Map<String, String> data;
   
   public DatabaseMemento(Map<String, String> dbData) {
       this.data = new HashMap<>(dbData);
   }
   
   public Map<String, String> getState() {
       return data;
   }
}

// Originator - The database whose state we want to save/restore
class Database {
   private Map<String, String> records;
   
   public Database() {
       records = new HashMap<>();
   }
   
   // Insert a record
   public void insert(String key, String value) {
       records.put(key, value);
       System.out.println("Inserted: " + key + " = " + value);
   }
   
   // Update a record
   public void update(String key, String value) {
       if (records.containsKey(key)) {
           records.put(key, value);
           System.out.println("Updated: " + key + " = " + value);
       } else {
           System.out.println("Key not found for update: " + key);
       }
   }
   
   // Delete a record
   public void remove(String key) {
       if (records.containsKey(key)) {
           records.remove(key);
           System.out.println("Deleted: " + key);
       } else {
           System.out.println("Key not found for deletion: " + key);
       }
   }
   
   // Create memento - Save current state
   public DatabaseMemento createMemento() {
       System.out.println("Creating database backup...");
       return new DatabaseMemento(records);
   }
   
   // Restore from memento - Rollback to saved state
   public void restoreFromMemento(DatabaseMemento memento) {
       records = new HashMap<>(memento.getState());
       System.out.println("Database restored from backup!");
   }
   
   // Display current database state
   public void displayRecords() {
       System.out.println("\n--- Current Database State ---");
       if (records.isEmpty()) {
           System.out.println("Database is empty");
       } else {
           for (Map.Entry<String, String> record : records.entrySet()) {
               System.out.println(record.getKey() + " = " + record.getValue());
           }
       }
       System.out.println("-----------------------------\n");
   }
}

// Caretaker - Manages the memento (transaction manager)
class TransactionManager {
   private DatabaseMemento backup;
   
   public TransactionManager() {
       backup = null;
   }
   
   // Begin transaction - create backup
   public void beginTransaction(Database db) {
       System.out.println("=== BEGIN TRANSACTION ===");
       backup = db.createMemento();
   }
   
   // Commit transaction - discard backup
   public void commitTransaction() {
       System.out.println("=== COMMIT TRANSACTION ===");
       if (backup != null) {
           backup = null;
       }
       System.out.println("Transaction committed successfully!");
   }
   
   // Rollback transaction - restore from backup
   public void rollbackTransaction(Database db) {
       System.out.println("=== ROLLBACK TRANSACTION ===");
       if (backup != null) {
           db.restoreFromMemento(backup);
           backup = null;
           System.out.println("Transaction rolled back!");
       } else {
           System.out.println("No backup available for rollback!");
       }
   }
}

public class MementoPattern {
   public static void main(String[] args) {
       Database db = new Database();
       TransactionManager txManager = new TransactionManager();
      
       //success scenario
       txManager.beginTransaction(db);
       db.insert("user1", "Aditya");
       db.insert("user2", "Rohit");
       txManager.commitTransaction();

       db.displayRecords();

       // Failed scenario
       txManager.beginTransaction(db);
       db.insert("user3", "Saurav");
       db.insert("user4", "Manish");

       db.displayRecords();
       
       // Some error -> Rollback
       System.out.println("ERROR: Something went wrong during transaction!");
       txManager.rollbackTransaction(db);
       
       db.displayRecords();
   }
}